File: loxide/src/table.rs
use std::ptr::{null_mut, NonNull};

use crate::{
    mem::{Gc, Greystack},
    obj::{Obj, ObjString},
    value::Value,
};

#[derive(Copy, Clone, Debug, PartialEq)]
pub struct ObjHash(pub u32);

impl ObjHash {
    pub const EMPTY_STR_HASH: Self = ObjHash(2166136261u32);

    pub fn hash_string(string: &str) -> ObjHash {
        if string.is_empty() {
            return Self::EMPTY_STR_HASH;
        }

        let string_bytes = string.as_bytes();
        let mut hash = 2166136261u32;

        for i in 0..string.len() {
            hash ^= string_bytes[i] as u32;
            hash = hash.overflowing_mul(16777619).0;
            // hash *= ;
        }

        ObjHash(hash)
    }
}

pub struct TableIter<'a> {
    table: &'a Table,
    index: usize,
}

impl<'a> Iterator for TableIter<'a> {
    type Item = &'a Entry;

    fn next(&mut self) -> Option<Self::Item> {
        if self.table.len == 0 {
            return None;
        }

        loop {
            if self.index >= self.table.cap as usize {
                return None;
            }

            let entry = unsafe { self.table.entries.add(self.index).as_ref().unwrap() };

            self.index += 1;

            // Skip uninitialized and tombstone entries
            if !entry.key.is_null() {
                return Some(entry);
            }
        }
    }
}

pub struct TableIterMut<'a> {
    table: &'a mut Table,
    index: usize,
}

impl<'a> Iterator for TableIterMut<'a> {
    type Item = &'a mut Entry;

    fn next(&mut self) -> Option<Self::Item> {
        if self.table.len == 0 {
            return None;
        }

        loop {
            if self.index >= self.table.cap as usize {
                return None;
            }

            let entry = unsafe { self.table.entries.add(self.index).as_mut().unwrap() };

            self.index += 1;

            // Skip uninitialized and tombstone entries
            if !entry.key.is_null() {
                return Some(entry);
            }
        }
    }
}

#[derive(Clone)]
pub struct Table {
    pub len: u32,
    pub cap: u32,
    pub entries: *mut Entry,
}

#[derive(Copy, Clone)]
pub struct Entry {
    pub key: *mut ObjString,
    pub value: Value,
}

impl std::fmt::Debug for Table {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Table")
            .field("len", &self.len)
            .field("cap", &self.cap)
            .field("entries", &self.entries)
            .finish()
    }
}

impl std::fmt::Debug for Entry {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        unsafe {
            f.debug_struct("Entry")
                .field(
                    "key",
                    &match self.key.as_ref() {
                        Some(obj_str) => obj_str.as_str(),
                        None => "null_key",
                    },
                )
                .field("(key_ptr)", &self.key)
                .field("value", &self.value)
                .finish()
        }
    }
}

impl Table {
    pub const TABLE_MAX_LOAD: f32 = 0.75;

    pub fn new() -> Self {
        Self {
            len: 0,
            cap: 0,
            entries: null_mut(),
        }
    }

    pub fn iter(&self) -> TableIter {
        TableIter {
            table: self,
            index: 0,
        }
    }

    pub fn iter_mut(&mut self) -> TableIterMut {
        TableIterMut {
            table: self,
            index: 0,
        }
    }

    fn entries_slice(&self) -> Option<&[Entry]> {
        // Safety:
        // `self.entries` layout is an array of `self.cap` length so its okay to create a slice of it
        if self.entries.is_null() {
            None
        } else {
            Some(unsafe { std::slice::from_raw_parts(self.entries, self.cap as usize) })
        }
    }

    pub fn add_all(&self, to: &mut Self) {
        match self.entries_slice() {
            Some(entries) => {
                for entry in entries.iter() {
                    match NonNull::new(entry.key) {
                        Some(key) => {
                            to.set(key, entry.value);
                        }
                        None => (),
                    }
                }
            }
            None => (),
        }
    }

    pub fn adjust_capacity(&mut self, new_cap: u32) {
        let mut entries = vec![Entry::uninitialized(); new_cap as usize];

        if !self.entries.is_null() {
            let mut new_len = 0;

            // Drop old
            // Safety:
            // We don't need to call drop on each Entry so Vec deallocating is just fine
            let old_entries =
                unsafe { Vec::from_raw_parts(self.entries, self.cap as usize, self.cap as usize) };

            for entry in old_entries.iter() {
                unsafe {
                    let key = match NonNull::new(entry.key) {
                        Some(key) => key,
                        None => continue,
                    };

                    let dest = Self::find_entry_from_ptr(entries.as_mut_ptr(), new_cap, key);
                    (*dest).key = key.as_ptr();
                    (*dest).value = entry.value;
                }
                new_len += 1;
            }

            self.len = new_len;
        }

        self.entries = entries.as_mut_ptr();
        self.cap = new_cap;

        let _ = entries.leak();
    }

    pub fn set(&mut self, key: NonNull<ObjString>, val: Value) -> bool {
        if self.len as f32 + 1.0 > self.cap as f32 * Self::TABLE_MAX_LOAD {
            let new_cap = if self.cap < 8 { 8 } else { self.cap * 2 };
            self.adjust_capacity(new_cap);
        }

        let entry = self.find_entry_mut(key);

        let is_new_key = entry.is_uninitialized();
        let should_increment_len = entry.is_uninitialized();

        entry.key = key.as_ptr();
        entry.value = val;

        if should_increment_len {
            self.len += 1;
        }

        is_new_key
    }

    pub fn delete(&mut self, key: NonNull<ObjString>) -> bool {
        if self.len == 0 {
            return false;
        }

        let entry = self.find_entry_mut(key);
        if entry.key.is_null() {
            return false;
        }

        // place tombstone
        entry.delete();
        true
    }

    pub fn get(&self, key: NonNull<ObjString>) -> Option<Value> {
        if self.len == 0 {
            return None;
        }

        let entry = self.find_entry(key);
        if entry.key.is_null() {
            return None;
        }

        Some(entry.value)
    }

    pub fn find_entry(&self, key: NonNull<ObjString>) -> &Entry {
        unsafe {
            Self::find_entry_from_ptr(self.entries, self.cap, key)
                .as_ref()
                .unwrap()
        }
    }

    pub fn find_entry_mut(&mut self, key: NonNull<ObjString>) -> &mut Entry {
        unsafe {
            Self::find_entry_from_ptr(self.entries, self.cap, key)
                .as_mut()
                .unwrap()
        }
    }

    pub fn find_string(&self, string: &str, hash: ObjHash) -> Option<Gc<ObjString>> {
        if self.len == 0 {
            return None;
        }

        let entries = match self.entries_slice() {
            Some(entries) => entries,
            None => return None,
        };

        let mut index = hash.0 & (self.cap - 1);

        unsafe {
            loop {
                let entry = entries[index as usize];
                if entry.key.is_null() {
                    if matches!(entry.value, Value::Nil) {
                        return None;
                    }
                } else if (*entry.key).len == string.len() as u32
                    && (*entry.key).hash == hash
                    && (*entry.key).as_str() == string
                {
                    return Some(Gc::new(NonNull::new_unchecked(entry.key)));
                }

                index = (index + 1) & (self.cap - 1);
            }
        }
    }

    fn find_entry_from_ptr(entries: *mut Entry, cap: u32, key: NonNull<ObjString>) -> *mut Entry {
        let mut index = unsafe { (*key.as_ptr()).hash.0 } & (cap - 1);
        let mut tombstone: *mut Entry = null_mut();

        loop {
            unsafe {
                let entry = entries.offset(index as isize);
                if (*entry).key.is_null() {
                    // It's a tombstone
                    if matches!((*entry).value, Value::Nil) {
                        return if !tombstone.is_null() {
                            tombstone
                        } else {
                            entry
                        };
                    }
                    tombstone = entry;
                } else if (*entry).key == key.as_ptr() {
                    return entry;
                }

                index = (index + 1) & (cap - 1);
            }
        }
    }

    pub fn free(table: &mut Table) {
        if table.entries.is_null() {
            return;
        }

        // free the buckets
        let _entries =
            unsafe { Vec::from_raw_parts(table.entries, table.len as usize, table.cap as usize) };

        table.len = 0;
        table.cap = 0;
        table.entries = null_mut();
    }

    pub fn mark(&self, greystack: &mut Greystack) {
        for entry in self.iter() {
            Obj::mark(entry.key.cast(), greystack);
            entry.value.mark(greystack);
        }
    }

    pub fn remove_white(&mut self) {
        for entry in self.iter_mut() {
            let is_marked = unsafe {
                // Safety:
                // TableIterMut skips over entries with null keys (unintialized and tombstoned entries)
                !entry
                    .key
                    .cast::<Obj>()
                    .as_ref()
                    .unwrap_unchecked()
                    .is_marked
            };

            if is_marked {
                entry.delete();
            }
        }
    }
}

impl Entry {
    fn delete(&mut self) {
        self.key = null_mut();
        // place tombstone
        self.value = Value::Bool(true);
    }

    fn uninitialized() -> Self {
        Self {
            key: null_mut(),
            value: Value::Nil,
        }
    }

    fn is_uninitialized(&self) -> bool {
        self.key.is_null() && self.value.is_nil()
    }

    fn is_tombstone(&self) -> bool {
        self.key.is_null() && matches!(self.value, Value::Bool(true))
    }
}

File: zlox/src/table.zig
const std = @import("std");
const Allocator = @import("std/mem").Allocator;
const Obj = @import("obj.zig");
const Value = @import("value.zig").Value;
const GC = @import("gc.zig");

const Table = @This();

count: u32 = 0,
cap: u32 = 0,
entries: ?[*]Entry = null,

const MAX_LOAD = 0.75;

pub const Entry = struct { key: ?*Obj.String, val: Value };

pub fn init() Table {
    return .{};
}

pub fn free(self: *Table, gc: *GC) void {
    if (self.entries) |entries| {
        gc.free(entries[0..self.cap]);
    }
    self.count = 0;
    self.cap = 0;
    self.entries = null;
}

pub fn hash_string(key: [*]const u8, len: u32) u32 {
    var hash: u32 = 2166136261;
    for (key[0..len]) |c| {
        hash ^= c;
        hash *= 16777619;
    }
    return hash;
}

pub fn get(self: *Table, key: *Obj.String) ?Value {
    if (self.count == 0) return null;

    const entry = self.find_entry(key) orelse return null;
    if (entry.key == null) return null;

    return entry.val;
}

/// Returns true if the key is new.
pub fn insert(self: *Table, gc: *GC, key: *Obj.String, val: Value) !bool {
    if (@intToFloat(f64, self.count + 1) > @intToFloat(f64, self.cap) * MAX_LOAD) {
        const new_cap = if (self.cap < 8) 8 else self.cap * 2;
        try self.adjust_capacity(gc, new_cap);
    }

    // Above if branch will always set entries.
    const entry = self.find_entry(key) orelse unreachable;
    const is_new_key = entry.key == null;
    if (is_new_key and entry.val.as_nil() != null) {
        self.count += 1;
    }

    entry.key = key;
    entry.val = val;
    return is_new_key;
}

pub fn add_all(self: *const Table, gc: *GC, to: *Table) !void {
    const entries = self.entries_slice() orelse return;
    for (entries) |entry| {
        if (entry.key == null) continue;
        try to.insert(gc, entry.key, entry.val);
    }
}

pub fn delete(self: *Table, key: *Obj.String) bool {
    if (self.count == 0) return false;
    var entry = self.find_entry(key) orelse return false;
    if (entry.key == null) return false;

    entry.key = null;
    entry.val = Value.boolean(true);
    return true;
}

pub fn find_string(self: *Table, chars: [*]const u8, len: u32, hash: u32) ?*Obj.String {
    if (self.count == 0) return null;
    const entries = self.entries_slice() orelse return null;
    var idx = hash % self.cap;

    var entry: *Entry = undefined;
    while (true) {
        entry = &entries[idx];
        if (entry.key) |key| {
            if (key.len == len and key.hash == hash and std.mem.eql(u8, key.chars[0..key.len], chars[0..len])) return key;
        } else if (entry.val.as_nil() != null) {
            return null;
        }

        idx = (idx + 1) % self.cap;
    }
}

fn find_entry(self: *Table, key: *Obj.String) ?*Entry {
    var entries = self.entries_slice() orelse return null;
    return find_entry_impl(entries, key);
}

fn find_entry_impl(entries: []Entry, key: *Obj.String) *Entry {
    const cap = entries.len;
    var idx = key.hash % cap;
    var tombstone: ?*Entry = null;
    while (true) {
        const entry = &entries[idx];
        if (entry.key == null) {
            if (entry.val.as_nil() != null) {
                if (tombstone) |ts| {
                    return ts;
                }
                return entry;
            }

            if (tombstone == null) {
                tombstone = entry;
            }
        } else if (entry.key == key) {
            return entry;
        }

        idx = (idx + 1) % cap;
    }
}

pub fn adjust_capacity(self: *Table, gc: *GC, new_cap: usize) !void {
    var new_entries = try gc.alloc(Entry, new_cap);
    for (new_entries) |*entry| {
        entry.key = null;
        entry.val = Value.nil();
    }

    var count: u32 = 0;
    if (self.entries_slice()) |entries| {
        for (entries) |entry| {
            const key = entry.key orelse continue;

            var dest = find_entry_impl(new_entries, key);
            dest.key = key;
            dest.val = entry.val;
            count += 1;
        }
        gc.free(entries);
    }

    self.entries = @ptrCast([*]Entry, new_entries);
    self.cap = @intCast(u32, new_cap);
    self.count = count;
}

pub fn entries_slice(self: *Table) ?[]Entry {
    if (self.entries) |entries| {
        return entries[0..self.cap];
    }

    return null;
}
